### ISSUE TITLE
Formalize HyperLogLog struct alignment for interoperability.

### ISSUE THREAD

**Issue by joshk0** [pipelinedb/pipelinedb#1460]
We want to insert HyperLogLog literals into our database and have been generating binary blobs from our HLL implementation (retailnext/hllpp) that match the struct format of our current build of PipelineDB. This struct format is a moving target. To allow it not to be, I'd like to propose that some variant of the struct alignment be formalized, and the fields changed to use explicit byte-width types, to support futuer interoperability. Alternatively, we could develop a separate, aligned export format and define a hll_import function that understands this format, and leave the core HyperLogLog struct unversioned and unaligned.

In other words, I want to be able to do `select hll_cardinality(hll_import('\xEXPORTBLOB'));` or `select hll_cardinality('\xPIPELINEHLLBLOB')` in psql and not have to worry that the blob format drifts between versions of Pipeline.

It seems like fixing an export format might be good for you in the long run especially because it would support backup / restore of HLLs between versions.

There's many ways to move forward here, would love to discuss and help formulate a good plan!

**Comment by usmanm** [comment]
Do you plan on only using imported HLLs for read queries?

**Comment by joshk0** [comment]
It's not terribly important. They can be write-once, as we intend to only do the ETL into PipelineDB once a day and the data is time-oriented, so once a given day is over its values should never change.

**Comment by joshk0** [comment]
Having studied this, is it true that if you revise `struct HyperLogLog`'s structure that all HLLs in a PipelineDB instance would be invalidated? Or is there some kind of migration scheme? Does pipeline-dump just output the struct blobs verbatim?

**Comment by usmanm** [comment]
Having them support updates is going to be tricky because that will require identical hashing of values in your Go library and PipelineDB. Reads should be fine though, since we'll just re-construct the HLL in our format.

Yes, all of the data would be invalid. `pipeline-dump` dumps the raw bytes. Adding versions to these data types was a thought we had in the beginning, but I feel like they're structure is fairly stable and will probably never change. At least not likely for HLL, Bloom, CMS, T-Digests. We might do more work on FSS, the implementation right now is very basic.

**Comment by usmanm** [comment]
I think for your use case, the easiest thing to do would be to create a `hll_import` function which converts the HLL from your representation to our internal representation.

**Comment by joshk0** [comment]
Right. We were looking not to have to modify PipelineDB source code for now -- at least not on our own. I think hll_import is fairly extraneous, because in the end that will take a byte array too, just of a different type. So it's either you version the format of that binary blob, or you version the real one and just insert it verbatim.

It's worth noting that at this point in time you are still vulnerable to compiler changes that cause different alignment decisions for the HyperLogLog struct on the same architecture simply because you haven't explicitly specified the alignment. Not to say that these would come up anytime soon, but it seems that if you are developing an enterprise product, you need a bulletproof way to transfer this data across architectures, especially because the hll type is part of what makes your product unique.

Low effort path is this IMHO.

- Define a canonical version of the HyperLogLog struct with explicit padding. Ideally this would be the one that is found most often in the wild.
- Fix `card` and `mlen` to use explicit uint64_t and uint32_t. This removes your reliance on long meaning 64-bit integer. Furthermore, these two variables have no business being negative, so they should be defined as unsigned.
- Prescribe the endianness of the numeric members of the struct, and use some macros to convert to and from the standard byte order that are no-ops where the native byte order is already correct. (As HyperLogLog.M is accessed byte-wise already, this shouldn't affect performance too much.)

Boom, now you have a type that is invariant across all platforms and provides a reference point for import/export. I'm happy to help with executing all of this if you're interested.

**Comment by usmanm** [comment]
Yea, an import function on our side or an export on your side is the same thing. Maybe exporting it in the our binary format is the easier option for you because it doesn't require writing any PG code.

Thanks for thinking about all of this man! I definitely agree with you that we should make these data structures more robust, and compiler changes shouldn't break anything. I'm going to create an issue for auditing all data structures and making sure the binary blobs are safe again compiler optimizations etc. Going to try to get that in before the next release.

**Comment by usmanm** [comment]
Merging with #1464
